keps/sig-api-machinery/20191210-consistent-reads-from-cache.md

```diff
@@ -0,0 +1,252 @@
+  - sig-scalability
+reviewers:
+  - TBD
+  - "@wojtek-t"
+  - "@jingyih"
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2019-12-10
+last-updated: 2019-12-10
+status: provisional
+see-also:
+replaces:
+superseded-by:
+---
+
+# Consistent Reads from Cache
+
+Kubernetes Get and List requests are guaranteed to be "consistent reads" if the
+`resourceVersion` parameter is not provided. Consistent reads are served from
+etcd using a "quorum read".
+
+But often the watch cache contains sufficiently up-to-date data to serve the
+read request, and could serve it far more efficiently.
+
+This KEP proposes a mechanism to serve most reads from the watch cache
+while still providing the same consistency guarantees as serving the
+read from etcd.
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+- [Proposal](#proposal)
+  - [Make the kube-apiserver aware of etcd's minor version](#make-the-kube-apiserver-aware-of-etcds-minor-version)
+  - [Leveraging the Progress Notify Mechanism](#leveraging-the-progress-notify-mechanism)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Pagination](#pagination)
+  - [Test Plan](#test-plan)
+  - [Rollout Plan](#rollout-plan)
+    - [Serving consistent reads from cache](#serving-consistent-reads-from-cache)
+    - [Reflectors](#reflectors)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Alternatives](#alternatives)
+- [Potential Future Improvements](#potential-future-improvements)
+<!-- /toc -->
+
+## Summary
+
+Consistent reads may be served from cache so long as:
+- A consistent (quorum) read is first made to etcd to get the latest "revision"
+- The data in the watch cache no older than the latest "revision" just from etcd
+
+etcd 3.4 supports progress events. [Progress event
+interval](https://github.com/etcd-io/etcd/blob/e6980b1f9fc008837abb9177e87e893e48d565c1/etcdserver/api/v3rpc/watch.go#L67)
+are sent by etcd automatically, and [progress
+notify](https://github.com/etcd-io/etcd/issues/9855) requests a progress event
+and was added specifically to make it easy to check if a cache that is updated
+via an etcd watch is up-to-date relative to some etcd revision).
+
+This KEP summarizes how we can take advantage of progress events efficiently
+determine how up-to-date kubernetes watch caches are then serve reads from the
+watch cache when they are sufficiently up-to-date.
+
+## Motivation
+
+Serving reads from the watch cache is more performant and scalable than reading
+them from etcd, deserializing them, converting them to the desired type and then
+garbage collecting all the objects that were allocated during the read.
+
+We will need to measure the impact to performance and scalability, but we have
+data and experience from prior improvements made by the watch cache that
+suggests there is significant opportunity here, and we would like to explore it.
```

Comment:
nit: I'm 100% convinced that this will help performance/scalability - issues like https://github.com/kubernetes/kubernetes/issues/86483 actually already prove it. We just need to measure by how much.

Reply:
I agree. I've rephrased to match.